File: sdk/flutter/iapstack/lib/src/retry_policy.dart
/// Retry settings for idempotent verification, restore, and entitlement calls.
final class IapStackRetryPolicy {
  /// Creates a bounded full-jitter exponential backoff policy.
  const IapStackRetryPolicy({
    this.maxAttempts = 3,
    this.baseDelay = const Duration(milliseconds: 250),
    this.maxDelay = const Duration(seconds: 2),
  });

  /// Total attempts, including the initial request.
  final int maxAttempts;

  /// Upper delay bound before the second attempt.
  final Duration baseDelay;

  /// Upper delay bound.
  final Duration maxDelay;

  /// Returns a full-jitter delay after a failed one-based [attempt].
  Duration delayAfter(int attempt, {required double randomValue}) {
    if (randomValue < 0 || randomValue > 1) {
      throw ArgumentError.value(
          randomValue, 'randomValue', 'must be between 0 and 1');
    }
    if (baseDelay == Duration.zero) {
      return Duration.zero;
    }
    final exponent = attempt - 1;
    final multiplier = 1 << exponent.clamp(0, 20);
    final microseconds = baseDelay.inMicroseconds * multiplier;
    final capped = microseconds.clamp(0, maxDelay.inMicroseconds);
    return Duration(microseconds: (capped * randomValue).round());
  }

  /// Throws [ArgumentError] when bounds are invalid.
  void validate() {
    if (maxAttempts < 1 || maxAttempts > 5) {
      throw ArgumentError.value(
          maxAttempts, 'maxAttempts', 'must be between 1 and 5');
    }
    if (baseDelay.isNegative || maxDelay.isNegative || baseDelay > maxDelay) {
      throw ArgumentError('retry delays must be non-negative and ordered');
    }
  }
}

File: sdk/react-native/src/retry_policy.ts
export class IapStackRetryPolicy {
  readonly maxAttempts: number;
  readonly baseDelayMs: number;
  readonly maxDelayMs: number;

  constructor(init: Partial<IapStackRetryPolicy> = {}) {
    this.maxAttempts = init.maxAttempts ?? 3;
    this.baseDelayMs = init.baseDelayMs ?? 250;
    this.maxDelayMs = init.maxDelayMs ?? 2000;
    this.validate();
  }

  delayAfter(attempt: number, randomValue: number): number {
    if (randomValue < 0 || randomValue > 1 || Number.isNaN(randomValue)) {
      throw new RangeError('randomValue must be between 0 and 1');
    }
    if (this.baseDelayMs <= 0) {
      return 0;
    }
    const exponent = attempt - 1;
    const multiplier = Math.min(1 << Math.max(0, Math.min(20, exponent)), 1_000_000);
    const capped = Math.min(this.baseDelayMs * multiplier, this.maxDelayMs);
    return Math.round(capped * randomValue);
  }

  validate(): void {
    if (!Number.isInteger(this.maxAttempts) || this.maxAttempts < 1 || this.maxAttempts > 5) {
      throw new TypeError('maxAttempts must be an integer between 1 and 5');
    }
    if (this.baseDelayMs < 0 || this.maxDelayMs < 0 || this.baseDelayMs > this.maxDelayMs) {
      throw new TypeError('retry delays must be non-negative and ordered');
    }
  }
}
